File: ds_arrays_02.js
class MyArray {
  constructor() {
    this.length = 0
    this.data = {}
  }
  //O(1)
  get(index) {
    return this.data[index]
  }
  //O(1)
  push(item) {
    this.data[this.length] = item
    this.length++
    return this.data
  }
  pop() {
    const lastItem = this.data[this.length - 1]
    delete this.data[this.length - 1]
    this.length--
    return lastItem
  }
  //O(n)
  deleteAtIndex(index) {
    const item = this.data[index]
    this.shiftItems(index)
    return item
  }
  //O(n)
  shiftItems(index) {
    for (let i = index; i < this.length - 1; i++) {
      this.data[i] = this.data[i + 1]
    }
    console.log('inside shift for deleteAtIndex', this.data[this.length - 1])
    delete this.data[this.length - 1]
    this.length--
  }
}

const myArray = new MyArray()
myArray.push('hi')
myArray.push('you')
myArray.push('!')
console.log('***after 3 push')
console.log(myArray)
console.log('***after 1 pop')
myArray.pop()
console.log(myArray)
console.log('***after deleteAtIndex(0)')
myArray.deleteAtIndex(0)
console.log(myArray)
myArray.push('are')
myArray.push('nice')
console.log(myArray)
myArray.shiftItems(0)
console.log(myArray)

File: ds_arrays_02.py
class Array:
  #constructor
  def __init__(self):
    self.length=0
    self.data={}

  def __str__(self):
    return str(self.__dict__)

  #O(n)
  def get(self,index):
    return self.data[index]
  
  #O(1)
  def push(self,item):
    self.data[self.length]=item
    self.length+=1

  def pop(self):
    lastitem = self.data[self.length-1]
    del self.data[self.length-1]
    self.length-=1
    return lastitem

  #O(n)
  def deleteAtIndex(self,index):
    deleteditem = self.data[index]
    for i in range(index,self.length-1):
      self.data[i] = self.data[i+1]

    del self.data[self.length-1]
    self.length-=1
    return deleteditem

myArray=Array()
myArray.push('hi')
myArray.push('you')
myArray.push('!')
('***after 3 push')
print(myArray)
print('***after 1 pop')
myArray.pop()
print(myArray)
print('***after deleteAtIndex(0)')
myArray.deleteAtIndex(0)
print(myArray)
myArray.push('are')
myArray.push('nice')
print(myArray)
myArray.deleteAtIndex(0)
print(myArray)

  